kratos/utilities/pointer_map_communicator.h

```diff
@@ -0,0 +1,623 @@
+
+    using DataVectorType = std::vector<TValueDataType>;
+
+    using DataVectorMapType = std::unordered_map<int, DataVectorType>;
+
+    ///@}
+    ///@name Life cycle
+    ///@{
+
+    /**
+     * @brief Construct a new Apply Proxy object
+     *
+     * @param rApplyFunctor                         Thread safe update lambda method
+     * @param rPointerCommunicator                  Map communicator
+     */
+    ApplyProxy(
+        const TApplyFunctor& rApplyFunctor,
+        TGPMapCommunicator& rPointerCommunicator)
+        : mCurrentRank(rPointerCommunicator.GetMyPID()),
+          mrApplyFunctor(rApplyFunctor),
+          mrPointerCommunicator(rPointerCommunicator)
+    {
+        KRATOS_TRY
+
+        if (rPointerCommunicator.IsDistributed()) {
+
+            KRATOS_DEBUG_ERROR_IF(OpenMPUtils::IsInParallel() != 0)
+                << "Constructing a proxy in a parallel "
+                   "region is not allowed.\n";
+
+            // resizing for OMP parallel loops
+            const int number_of_threads = OpenMPUtils::GetNumThreads();
+            mNonLocalGlobalPointerMapsVector.resize(number_of_threads);
+            mNonLocalDataValueMapsVector.resize(number_of_threads);
+
+            this->mUpdateMethod = &ProxyType::AssignLocalAndRemoteData;
+        } else {
+            this->mUpdateMethod = &ProxyType::AssignLocalData;
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    ///@}
+    ///@name Public operations
+    ///@{
+
+    /**
+     * @brief Assigns values with given values map
+     *
+     * This method assigns local gps using mrApplyFunctor instantly. Values
+     * belonging to remote gps are stored in the mrNonLocalGlobalPointerMapsVector and mrNonLocalDataValueMapsVector.
+     *
+     * In the case of serial job, this method calls AssignLocalData, where no checks are done
+     * to ensure gps are local because all gps are local.
+     *
+     * In case of distributed job, this method calls AssignLocalAndRemoteData where checks are
+     * performed to identify gps are local or non-local, based on that mrApplyFunctor
+     * methods are called or values corresponding to remote gps are stored.
+     *
+     * This does not have additional cost in serial run (except for function pointer call)
+     *
+     * @see GlobalPointerMapCommunicator
+     *
+     * @param rGPDataMap        Input values map (key: GlobalPointer<TPointerDataType>, value: TValueDataType)
+     */
+    void Assign(const TGPDataMap& rGPDataMap)
+    {
+        KRATOS_TRY
+
+        KRATOS_DEBUG_ERROR_IF(OpenMPUtils::IsInParallel() != 0)
+            << "Assigning map of values in a parallel region is not allowed.\n";
+
+        // get gp vector for parallel omp run
+        const auto& gps = TGPMapCommunicator::GetKeys(rGPDataMap);
+
+        // running this in parallel assuming mrApplyFunctor is thread safe
+        IndexPartition<int>(gps.size()).for_each([&](const int Index) {
+            auto p_itr = rGPDataMap.find(gps[Index]);
```

Comment:
do we still have a map? i thought it is now a vector ... and in any case hy do you access by index and not directly by iterator? note that access by index has an overhead

Reply:
The maps are supported so that the gp_list is unique so we can do OMP parallelization. I removed the vector support because with the current design the gp_list uniqueness is not ensured always. if we are to support vectors then we can not utilize OMP parallelism in serial or distributed cases... :/